picard/tagger.py

```diff
@@ -346,105 +363,152 @@ def move_files(self, files, target):
             self.cluster(files)
 
     def add_files(self, filenames, target=None):
+        self._add_files([self._add_files_from_directory("", filenames)], target=target)
+
+    def _add_files(self, list_of_filenames, target=None):
         """Add files to the tagger."""
-        ignoreregex = None
-        pattern = config.setting['ignore_regex']
-        if pattern:
-            ignoreregex = re.compile(pattern)
-        ignore_hidden = config.setting["ignore_hidden_files"]
-        new_files = []
+        number_of_files = sum([len(x) for x in list_of_filenames])
+        number_of_directories = len(list_of_filenames)
+        log.debug("Adding a total of %d files in %d directories",
+                number_of_files, number_of_directories)
+        self.window.set_statusbar_message(
+            N_("Adding a total of %(files)d files in %(directories)d directories ..."),
+            {'files': number_of_files, 'directories': number_of_directories},
+            echo=None
+        )
+        for filenames in list_of_filenames:
+            thread.run_task(partial(self._open_files, filenames),
+                    partial(self._open_files_finished, target),
+                    priority=2,
+                    thread_pool=self.load_thread_pool)
+
+    def _open_files(self, filenames):
+        new_files = {}
         for filename in filenames:
             filename = os.path.normpath(os.path.realpath(filename))
-            if ignore_hidden and is_hidden(filename):
-                log.debug("File ignored (hidden): %r" % (filename))
-                continue
-            if ignoreregex is not None and ignoreregex.search(filename):
-                log.info("File ignored (matching %r): %r" % (pattern, filename))
-                continue
             if filename not in self.files:
                 file = open_file(filename)
                 if file:
-                    self.files[filename] = file
-                    new_files.append(file)
-        if new_files:
+                    new_files[filename] = file
+                # Allow UI to process events to keep it responsive
+                QtCore.QCoreApplication.processEvents()
+        return new_files
+
+    def _open_files_finished(self, target=None, result=None, error=None):
+        if result:
+            new_files = []
+            for filename in sorted(result):
```

Comment:
Yes - it is necessary - files are not necessarily provided in track order - so what this does is display the tracks in correct sequence if they have track numbers (which in most cases will be true).